src/box/lua/key_def.c

```diff
@@ -99,9 +99,19 @@ luaT_push_key_def_parts(struct lua_State *L, const struct key_def *key_def)
 		}
 
 		if (part->coll_id != COLL_NONE) {
+			const char *name = NULL;
 			struct coll_id *coll_id = coll_by_id(part->coll_id);
-			assert(coll_id != NULL);
-			lua_pushstring(L, coll_id->name);
+			if (coll_id != NULL) {
+				/*
+				 * It is possible that the original collation
+				 * with `id == part->coll_id` was replaced by
+				 * the new one with the same id, but different
+				 * functional parts.
+				 */
+				if (coll_id_match_coll(coll_id, part->coll))
+					name = coll_id->name;
```

Comment:
`coll_new` reuses collation with the same fingerprint so you could simplify this check:

```C
if (coll_id != NULL && coll_id->coll == part->coll)
        name = coll_id->coll;
else
        name = "<deleted>";
```

Reply:
Done.